File: tests/java/org/elastos/hive/DatabaseServiceTest.java
package org.elastos.hive;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.JsonNodeFactory;
import com.fasterxml.jackson.databind.node.ObjectNode;
import org.elastos.hive.config.TestData;
import org.elastos.hive.service.DatabaseService;
import org.elastos.hive.vault.database.*;
import org.junit.jupiter.api.*;

import java.util.*;

@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
class DatabaseServiceTest {
	private static final String COLLECTION_NAME = "works";

	private static DatabaseService databaseService;

	@BeforeAll public static void setUp() {
		Assertions.assertDoesNotThrow(()->databaseService = TestData.getInstance()
				.newVault().getDatabaseService());
	}

	@Test @Order(1) void testCreateCollection() {
		Assertions.assertDoesNotThrow(()->{
			databaseService.createCollection(COLLECTION_NAME).get();
		});
	}

	@Test @Order(2) void testInsertOne() {
		Assertions.assertDoesNotThrow(()->{
			ObjectNode docNode = JsonNodeFactory.instance.objectNode();
			docNode.put("author", "john doe1");
			docNode.put("title", "Eve for Dummies1");
			Assertions.assertNotNull(databaseService.insertOne(COLLECTION_NAME, docNode,
					new InsertOptions().bypassDocumentValidation(false)).get());
		});
	}

	@Test @Order(3) void testInsertMany() {
		Assertions.assertDoesNotThrow(()->{
			List<JsonNode> nodes = new ArrayList<>();

			ObjectNode doc = JsonNodeFactory.instance.objectNode();
			doc.put("author", "john doe2");
			doc.put("title", "Eve for Dummies2");
			nodes.add(doc);

			doc.removeAll();

			doc = JsonNodeFactory.instance.objectNode();
			doc.put("author", "john doe3");
			doc.put("title", "Eve for Dummies3");
			nodes.add(doc);

			Assertions.assertNotNull(databaseService.insertMany(COLLECTION_NAME, nodes,
					new InsertOptions().bypassDocumentValidation(false).ordered(true)).get());
		});
	}

	@Test @Order(4) void testFindOne() {
		Assertions.assertDoesNotThrow(()->{
			ObjectNode query = JsonNodeFactory.instance.objectNode();
			query.put("author", "john doe1");
			Assertions.assertNotNull(databaseService.findOne(COLLECTION_NAME, query,
					new FindOptions().setSkip(0).setLimit(0)).get());
		});
	}

	@Test @Order(5) void testFindMany() {
		Assertions.assertDoesNotThrow(()->{
			ObjectNode query = JsonNodeFactory.instance.objectNode();
			query.put("author", "john doe1");
			List<JsonNode> docs = databaseService.findMany(COLLECTION_NAME, query,
					new FindOptions().setSkip(0).setLimit(0)).get();
		});
	}

	@Test @Order(6) void testQuery() {
		Assertions.assertDoesNotThrow(()->{
			ObjectNode query = JsonNodeFactory.instance.objectNode();
			query.put("author", "john doe1");
			Assertions.assertNotNull(databaseService.query(COLLECTION_NAME, query, null).get());
		});
	}

	@Test @Order(6) void testQueryWithOptions() {
		Assertions.assertDoesNotThrow(()->{
			ObjectNode query = JsonNodeFactory.instance.objectNode();
			query.put("author", "john doe1");
			QueryOptions options = new QueryOptions().setSort(new AscendingSortItem("_id"));
			Assertions.assertNotNull(databaseService.query(COLLECTION_NAME, query, options).get());
		});
	}

	@Test @Order(7) void testCountDoc() {
		Assertions.assertDoesNotThrow(()->{
			ObjectNode filter = JsonNodeFactory.instance.objectNode();
			filter.put("author", "john doe1");
			Assertions.assertNotNull(databaseService.countDocuments(COLLECTION_NAME, filter,
					new CountOptions().setLimit(1L).setSkip(0L).setMaxTimeMS(1000000000L)).get());
		});
	}

	@Disabled
	@Test @Order(8) void testUpdateOne() {
		Assertions.assertDoesNotThrow(()->{
			ObjectNode filter = JsonNodeFactory.instance.objectNode();
			filter.put("author", "john doe1");
			ObjectNode doc = JsonNodeFactory.instance.objectNode();
			doc.put("author", "john doe1");
			doc.put("title", "Eve for Dummies1_1");
			ObjectNode update = JsonNodeFactory.instance.objectNode();
			update.put("$set", doc);
			Assertions.assertNotNull(databaseService.updateOne(COLLECTION_NAME, filter, update,
					new UpdateOptions().setBypassDocumentValidation(false).setUpsert(true)).get());
		});
	}

	@Test @Order(9) void testUpdateMany() {
		Assertions.assertDoesNotThrow(()->{
			ObjectNode filter = JsonNodeFactory.instance.objectNode();
			filter.put("author", "john doe1");
			ObjectNode doc = JsonNodeFactory.instance.objectNode();
			doc.put("author", "john doe1");
			doc.put("title", "Eve for Dummies1_2");
			ObjectNode update = JsonNodeFactory.instance.objectNode();
			update.put("$set", doc);
			Assertions.assertNotNull(databaseService.updateMany(COLLECTION_NAME, filter, update,
					new UpdateOptions().setBypassDocumentValidation(false).setUpsert(true)).get());
		});
	}

	@Disabled
	@Test @Order(10) void testDeleteOne() {
		Assertions.assertDoesNotThrow(()->{
			ObjectNode filter = JsonNodeFactory.instance.objectNode();
			filter.put("author", "john doe2");
			databaseService.deleteOne(COLLECTION_NAME, filter).get();
		});
	}

	@Test @Order(11) void testDeleteMany() {
		Assertions.assertDoesNotThrow(()->{
			ObjectNode filter = JsonNodeFactory.instance.objectNode();
			filter.put("author", "john doe2");
			databaseService.deleteMany(COLLECTION_NAME, filter).get();
		});
	}

	@Test @Order(12) void testDeleteCollection() {
		Assertions.assertDoesNotThrow(()->
				databaseService.deleteCollection(COLLECTION_NAME).get());
	}
}

File: tests/src/database/databaseservice.test.ts
import { Vault, VaultSubscription, DatabaseService, AlreadyExistsException, InsertOptions, FindOptions, CountOptions, UpdateOptions, NotFoundException, QueryOptions, AscendingSortItem } from "../../../src";
import { TestData } from "../config/testdata";

describe("test database services", () => {
   
    let testData: TestData;
    let vaultSubscription: VaultSubscription;
    let vault: Vault;
    let databaseService: DatabaseService;

    let COLLECTION_NAME = "works";
    let COLLECTION_NAME_NOT_EXIST = "works_not_exists";

    beforeAll(async () => {

        let testData = await TestData.getInstance("databaseservice.tests");

        vaultSubscription = new VaultSubscription(
            testData.getAppContext(),
            testData.getProviderAddress());
        
        try {
            await vaultSubscription.subscribe();
        } catch (e){
            console.log("vault is already subscribed");
        }

        vault = new Vault(
            testData.getAppContext(),
            testData.getProviderAddress());
        
        databaseService = vault.getDatabaseService();
    });

    /*
    beforeEach(() => {
        console.log("***** Running " + expect.getState().currentTestName + " *****");
    });

    afterEach(() => {
        console.log("***** End of " + expect.getState().currentTestName + " *****");
    });
    */

    test("testCreateCollection", async () => {
        await deleteCollectionAnyway();
        await expect(databaseService.createCollection(COLLECTION_NAME)).resolves.not.toThrow();
    });

    test("testCreateCollection4AlreadyExistsException", async () => {
        let collectionName = `collection_${Date.now().toString()}`; 
        await databaseService.createCollection(collectionName);
        await expect(databaseService.createCollection(collectionName)).rejects.toThrow(AlreadyExistsException);
        await databaseService.deleteCollection(collectionName);
    });   

    test("testInsertOne", async () => {
        let docNode = {"author": "john doe1", "title": "Eve for Dummies1"};
        let result = await databaseService.insertOne(COLLECTION_NAME, docNode, new InsertOptions(false, false, true));
        expect(result).not.toBeNull();
        expect(result.getInsertedIds().length).toEqual(1);
    });

    test("testInsertOne4NotFoundException", async () => {
        let docNode = {"author": "john doe1", "title": "Eve for Dummies1"};
        await expect(databaseService.insertOne(COLLECTION_NAME_NOT_EXIST, docNode, new InsertOptions(false, false)))
            .rejects.toThrow(NotFoundException);
 	});

     test("testInsertMany", async () => {
        let nodes = [{"author": "john doe2", "title": "Eve for Dummies2"},
                     {"author": "john doe3", "title": "Eve for Dummies3"}];
        let result = await databaseService.insertMany(COLLECTION_NAME, nodes, new InsertOptions(false, true)); 
        expect(result).not.toBeNull();
        expect(result.getInsertedIds().length).toEqual(2);
    });

    test("testInsertMany4NotFoundException", async () => {
        let nodes = [{"author": "john doe2", "title": "Eve for Dummies2"},
                     {"author": "john doe3", "title": "Eve for Dummies3"}];
        await expect(databaseService.insertMany(COLLECTION_NAME_NOT_EXIST, nodes, new InsertOptions(false, false))).rejects.toThrow(NotFoundException); 
    });

    test("testFindOne", async () => {
        //setup
        let collectionName = TestData.getUniqueName("testFindOne");
        await databaseService.createCollection(collectionName);
        let nodes = {"author": "john doe1", "title": "Eve for Dummies1"};
        await expect(databaseService.insertOne(collectionName, nodes)).resolves.not.toThrow();

        let query = {"author": "john doe1"};
        let result = await databaseService.findOne(collectionName, query, new FindOptions());
        expect(result).not.toBeNull();
    });

    test("testFindOne4NotFoundException", async () => {
		let query = {"author": "john doe1"};
        await expect(databaseService.findOne(TestData.getUniqueName("testFindOne4NotFoundException"), query)).rejects.toThrow(NotFoundException);
	});
   
    test("testFindMany", async () => {
        let query = {"author": "john doe1"};
        await expect(databaseService.findMany(COLLECTION_NAME, query)).resolves.not.toThrow();
    });

    test("testFindMany4NotFoundException", async () => {
        let query = {"author": "john doe1"};
        await expect(databaseService.findMany(COLLECTION_NAME_NOT_EXIST, query)).rejects.toThrow(NotFoundException);
    });

    test("testQuery", async () => {
        let query = {"author": "john doe1"};
        await expect(databaseService.query(COLLECTION_NAME, query, null)).resolves.not.toBeNull();
    });

    test("testQuery4NotFoundException", async () => {
        let query = {"author": "john doe1"};
        await expect(databaseService.query(COLLECTION_NAME_NOT_EXIST, query)).rejects.toThrow(NotFoundException);
    });

    test("testQueryWithOptions", async () => {
        let query = {"author": "john doe1"};
        let options: QueryOptions = new QueryOptions();
        options.sort = [new AscendingSortItem("_id")];
        await expect(databaseService.query(COLLECTION_NAME, query, options)).resolves.not.toBeNull();
    });

    test("testCountDoc", async () => {
        let filter = { "author": "john doe1" };
        let countOptions = new CountOptions();
        countOptions.skip = 0;
        countOptions.limit = 1;
        countOptions.maxTimeMS = 1000000000;
        await expect(databaseService.countDocuments(COLLECTION_NAME, filter, countOptions)).resolves.not.toBeNull();
    });

    test("testCountDoc4NotFoundException", async () => {
        let filter = { "author": "john doe1" };
        let countOptions = new CountOptions();
        countOptions.skip = 0;
        countOptions.limit = 1;
        countOptions.maxTimeMS = 1000000000;
        await expect(databaseService.countDocuments(COLLECTION_NAME_NOT_EXIST, filter, countOptions)).rejects.toThrow(NotFoundException);
    });

    test("testUpdateOne", async () => {
        let filter = { "author": "john doe1" };
        let docNode = { "author": "john doe1", "title": "Eve for Dummies1_1" };
        let updateNode = { "$set": docNode };
        await expect(databaseService.updateOne(COLLECTION_NAME, filter, updateNode, new UpdateOptions(false, true))).resolves.not.toBeNull();
    });

    test("testUpdateInsertIfNotExists", async () => {
        /** The inserted document:
        {
            "_id": ObjectId("6241471ab042663cc9f179e7"),
            "author": "john doe4",
            "title": "title": "Eve for Dummies4"
        }
        */
        let filter = { "author": "john doe4" };
        let updateNode = { "$setOnInsert": { "title": "Eve for Dummies4" } };
        await expect(databaseService.updateOne(COLLECTION_NAME, filter, updateNode, new UpdateOptions(true, true))).resolves.not.toBeNull();
    });

    test("testUpdateOneNoOptions", async () => {
        let filter = { "author": "john doe1" };
        let docNode = { "author": "john doe1", "title": "Eve for Dummies1_1" };
        let updateNode = { "$set": docNode };
        await expect(databaseService.updateOne(COLLECTION_NAME, filter, updateNode)).resolves.not.toBeNull();
    });

    test("testUpdateOne4NotFoundException", async () => {
        let filter = { "author": "john doe1" };
        let docNode = { "author": "john doe1", "title": "Eve for Dummies1_1" };
        let updateNode = { "$set": docNode };
        await expect(databaseService.updateOne(COLLECTION_NAME_NOT_EXIST, filter, updateNode, new UpdateOptions(false, true))).rejects.toThrow(NotFoundException);
    });

    test("testUpdateMany", async () => {
        let filter = { "author": "john doe1" };
        let docNode = { "author": "john doe1", "title": "Eve for Dummies1_2" };
        let updateNode = { "$set": docNode };
        await expect(databaseService.updateMany(COLLECTION_NAME, filter, updateNode, new UpdateOptions(false, true))).resolves.not.toBeNull();
    });

    test("testUpdateMany4NotFoundException", async () => {
        let filter = { "author": "john doe1" };
        let docNode = { "author": "john doe1", "title": "Eve for Dummies1_2" };
        let updateNode = { "$set": docNode };
        await expect(databaseService.updateMany(COLLECTION_NAME_NOT_EXIST, filter, updateNode, new UpdateOptions(false, true))).rejects.toThrow(NotFoundException);
    });

    test("testDeleteOne", async () => {
        await expect(databaseService.deleteOne(COLLECTION_NAME, { "author": "john doe2" })).resolves.not.toThrow();
    });

    test("testDeleteOne4NotFoundException", async () => {
        await expect(databaseService.deleteOne(COLLECTION_NAME_NOT_EXIST, { "author": "john doe2" })).rejects.toThrow(NotFoundException);
    });

    test("testDeleteMany", async () => {
        await expect(databaseService.deleteMany(COLLECTION_NAME, { "author": "john doe2" })).resolves.not.toThrow();
    });

    test("testDeleteMany4NotFoundException", async () => {
        await expect(databaseService.deleteMany(COLLECTION_NAME_NOT_EXIST, { "author": "john doe2" })).rejects.toThrow(NotFoundException);
    });

    test("testDeleteCollection", async () => {
        await expect(databaseService.deleteCollection(COLLECTION_NAME)).resolves.not.toThrow();
    });

    async function deleteCollectionAnyway() {
        try {
            await databaseService.deleteCollection(COLLECTION_NAME);
        } catch (e) {
            // No error
        }
    }
});